File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/dialog/GPSDiagnosticDialog.java
package com.byagowi.persiancalendar.view.dialog;

import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.location.LocationManager;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.provider.Settings;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.Toast;

import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.databinding.DialogAccessBinding;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;

public class GPSDiagnosticDialog extends DialogFragment {

    public static boolean needsDiagnostic(Context context) {
        try {
            LocationManager gps = (LocationManager)
                    context.getSystemService(Context.LOCATION_SERVICE);
            ConnectivityManager connectivityManager = (ConnectivityManager)
                    context.getSystemService(Context.CONNECTIVITY_SERVICE);

            NetworkInfo info = null;
            if (connectivityManager != null) info = connectivityManager.getActiveNetworkInfo();

            boolean gpsEnabled = false;

            if (gps != null) {
                try {
                    gpsEnabled = gps.isProviderEnabled(LocationManager.GPS_PROVIDER);
                } catch (Exception ignored) {
                }
            }

            return !gpsEnabled || info == null;
        } catch (Exception e) {
            // Do whatever we were doing till now
            return false;
        }
    }

    // This is a workaround for the strange behavior of onCreateView (which doesn't show dialog's layout)
    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Context context = getContext();
        if (context == null) return null;

        AlertDialog.Builder dialogBuilder = new AlertDialog.Builder(context);
        DialogAccessBinding binding = DataBindingUtil.inflate(LayoutInflater.from(context),
                R.layout.dialog_access, null, false);
        dialogBuilder.setView(binding.getRoot());

        boolean gpsEnabled = false;
        NetworkInfo info = null;
        try {
            // check whether gps provider and network providers are enabled or not
            LocationManager gps = (LocationManager)
                    context.getSystemService(Context.LOCATION_SERVICE);
            ConnectivityManager connectivityManager = (ConnectivityManager)
                    context.getSystemService(Context.CONNECTIVITY_SERVICE);

            if (connectivityManager != null) {
                info = connectivityManager.getActiveNetworkInfo();
            }
            if (gps != null) {
                try {
                    gpsEnabled = gps.isProviderEnabled(LocationManager.GPS_PROVIDER);
                } catch (Exception ignored) {
                }
            }
        } catch (Exception ignore) {
        }

        if (!gpsEnabled && info == null) {
            Toast.makeText(context, R.string.internet_location_enable, Toast.LENGTH_SHORT).show();
            binding.dialogButtonGPS.setVisibility(View.VISIBLE);
            binding.dialogButtonWiFi.setVisibility(View.VISIBLE);
            binding.dialogButtonGPRS.setVisibility(View.VISIBLE);
        } else if (!gpsEnabled) {
            Toast.makeText(context, R.string.location_enable, Toast.LENGTH_SHORT).show();
            binding.dialogButtonGPRS.setVisibility(View.GONE);
            binding.dialogButtonWiFi.setVisibility(View.GONE);
        } else if (info == null) {
            Toast.makeText(context, R.string.internet_enable, Toast.LENGTH_SHORT).show();
            binding.dialogButtonGPS.setVisibility(View.GONE);
        }

        binding.dialogButtonGPS.setOnClickListener(v -> {
            Context ctx = getContext();
            if (ctx != null) {
                ctx.startActivity(new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS));
            }
            dismiss();
            // get gps
        });

        binding.dialogButtonWiFi.setOnClickListener(v -> {
            Context ctx = getContext();
            if (ctx != null) {
                ctx.startActivity(new Intent(Settings.ACTION_WIFI_SETTINGS));
            }
            dismiss();
            // get wifi
        });

        binding.dialogButtonGPRS.setOnClickListener(v -> {
            Context ctx = getContext();
            if (ctx != null) {
                ctx.startActivity(new Intent(Settings.ACTION_WIRELESS_SETTINGS));
            }
            dismiss();
            // get gprs
        });

        binding.dialogButtonExit.setOnClickListener(v -> {
            dismiss();
            // exit
        });

        return dialogBuilder.create();
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/view/dialog/GPSDiagnosticDialog.kt
package com.byagowi.persiancalendar.view.dialog

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.location.LocationManager
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.os.Bundle
import android.provider.Settings
import android.view.LayoutInflater
import android.view.View
import android.widget.Toast
import androidx.annotation.NonNull
import androidx.appcompat.app.AlertDialog
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.DialogFragment
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.databinding.DialogAccessBinding

class GPSDiagnosticDialog : DialogFragment() {

  // This is a workaround for the strange behavior of onCreateView (which doesn't show dialog's layout)
  @NonNull
  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
    val context = context ?: return Dialog(null)

    val dialogBuilder = AlertDialog.Builder(context)

    val binding = DataBindingUtil.inflate<DialogAccessBinding>(LayoutInflater.from(getContext()),
        R.layout.dialog_access, null, false)
    dialogBuilder.setView(binding.root)

    // check whether gps provider and network providers are enabled or not
    val gps = context.getSystemService(Context.LOCATION_SERVICE) as LocationManager?
    val connectivityManager = context.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager?
    val info: NetworkInfo? = connectivityManager?.activeNetworkInfo

    var gpsEnabled = false

    try {
      gpsEnabled = gps?.isProviderEnabled(LocationManager.GPS_PROVIDER) ?: false
    } catch (ignored: Exception) {
    }

    binding.dialogButtonGPS.setOnClickListener {
      val myIntent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
      activity?.startActivity(myIntent)
      dialog.dismiss()
      // get gps
    }

    binding.dialogButtonWiFi.setOnClickListener {
      val myIntent = Intent(Settings.ACTION_WIFI_SETTINGS)
      activity?.startActivity(myIntent)
      dialog.dismiss()
      // get wifi
    }

    binding.dialogButtonGPRS.setOnClickListener {
      val myIntent = Intent(Settings.ACTION_WIRELESS_SETTINGS)
      activity?.startActivity(myIntent)
      dialog.dismiss()
      // get gprs
    }

    binding.dialogButtonExit.setOnClickListener {
      dialog.dismiss()
      // exit
    }

    if (!gpsEnabled && info == null) {
      Toast.makeText(activity, R.string.internet_location_enable, Toast.LENGTH_SHORT).show()
      binding.dialogButtonGPS.visibility = View.VISIBLE
      binding.dialogButtonWiFi.visibility = View.VISIBLE
      binding.dialogButtonGPRS.visibility = View.VISIBLE
    } else if (!gpsEnabled) {
      Toast.makeText(activity, R.string.location_enable, Toast.LENGTH_SHORT).show()
      binding.dialogButtonGPRS.visibility = View.GONE
      binding.dialogButtonWiFi.visibility = View.GONE
    } else if (info == null) {
      Toast.makeText(activity, R.string.internet_enable, Toast.LENGTH_SHORT).show()
      binding.dialogButtonGPS.visibility = View.GONE
    }

    isCancelable = true

    return dialogBuilder.create()
  }
}